File: tenet-0/shared/bus-go/approvals.go
package bus

import (
	"context"
	"encoding/json"
	"fmt"
	"time"
)

// Approvals returns the approvals helper for this Bus. Per-action and blanket
// authorization helpers are President-only — non-President credentials get
// ErrNamespaceViolation back from the underlying Publish.
func (b *Bus) Approvals() *Approvals {
	return &Approvals{bus: b}
}

// Approvals wraps the publish_event SP for the three President approval
// event types. Non-President callers may still use these for completeness
// (testing, simulating), but the bus enforces namespace authorization.
type Approvals struct {
	bus *Bus
}

// GrantPerAction publishes a `president.approved` event scoped to a single
// target event. Returns the approval event ID, which the requesting
// department uses with bus.WithApproval(...) on its actual publish.
//
// expiresIn is the validity window from now. If zero, the SP applies its
// default (10 minutes) — pass an explicit duration to override.
func (a *Approvals) GrantPerAction(ctx context.Context, targetEventID, scope string, expiresIn time.Duration, reason string) (string, error) {
	payload := map[string]any{
		"approves_event_id": targetEventID,
		"scope":             scope,
		"reason":            reason,
	}
	if expiresIn > 0 {
		payload["expires_at"] = time.Now().Add(expiresIn).UTC().Format(time.RFC3339)
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("approvals: marshal: %w", err)
	}
	return a.bus.Publish(ctx, "president.approved", body)
}

// GrantBlanket publishes a `president.authorization.granted` event covering
// every future event in the named category. Pass a zero expiresAt for an
// indefinite (revoke-only) authorization.
//
// constraints is category-specific (e.g., {"max_amount_cents": 10000} for
// `routine.finance.small_refund`). May be nil.
func (a *Approvals) GrantBlanket(ctx context.Context, category string, constraints map[string]any, expiresAt time.Time, reason string) (string, error) {
	payload := map[string]any{
		"category": category,
		"reason":   reason,
	}
	if !expiresAt.IsZero() {
		payload["expires_at"] = expiresAt.UTC().Format(time.RFC3339)
	}
	if constraints != nil {
		payload["constraints"] = constraints
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("approvals: marshal: %w", err)
	}
	return a.bus.Publish(ctx, "president.authorization.granted", body)
}

// Revoke publishes a `president.authorization.revoked` event for a category.
// All currently active blanket authorizations matching the category are
// marked revoked atomically by the SP.
//
// approvalEventID is recorded in the payload for audit traceability — it does
// not need to be the original grant's ID; pass any human-meaningful reference.
func (a *Approvals) Revoke(ctx context.Context, approvalEventID, reason string) error {
	// Look up the category from the approvals_active row to make Revoke
	// callable with just the original approval event ID.
	var category *string
	err := a.bus.pool.QueryRow(ctx,
		`SELECT category FROM approvals_active
		  WHERE approval_event_id = $1 AND kind = 'blanket' LIMIT 1`,
		approvalEventID,
	).Scan(&category)
	if err != nil {
		return fmt.Errorf("approvals: lookup category: %w", err)
	}
	if category == nil {
		return fmt.Errorf("approvals: blanket approval %s has no category", approvalEventID)
	}

	payload, err := json.Marshal(map[string]any{
		"category":             *category,
		"revoked_approval_id":  approvalEventID,
		"reason":               reason,
	})
	if err != nil {
		return fmt.Errorf("approvals: marshal: %w", err)
	}
	_, err = a.bus.Publish(ctx, "president.authorization.revoked", payload)
	return err
}

File: tenet-0/shared/bus-ts/src/approvals.ts
import type { Bus } from "./bus.js";

export interface PerActionOptions {
  // ID of the pending request event this approval targets (the one the
  // requesting department published as `<dept>.approval.requested`).
  targetEventId: string;
  // Free-form scope descriptor recorded in the payload for audit.
  scope: string;
  // Validity window from now. If omitted, the SP applies its default (10 min).
  expiresInMs?: number;
  reason?: string;
}

export interface BlanketOptions {
  category: string;
  // Category-specific shape (e.g. {max_amount_cents: 10000}). Optional.
  constraints?: Record<string, unknown>;
  expiresAt?: Date;
  reason?: string;
}

// Approvals is the President-side API for granting and revoking approvals.
// The library does not enforce "president-only" — that's a stored-procedure
// namespace rule; non-president callers receive ErrNamespaceViolation from
// the underlying Publish.
export class Approvals {
  private readonly bus: Bus;

  constructor(bus: Bus) {
    this.bus = bus;
  }

  async grantPerAction(opts: PerActionOptions): Promise<string> {
    const payload: Record<string, unknown> = {
      approves_event_id: opts.targetEventId,
      scope: opts.scope,
      reason: opts.reason ?? "",
    };
    if (opts.expiresInMs && opts.expiresInMs > 0) {
      payload.expires_at = new Date(
        Date.now() + opts.expiresInMs,
      ).toISOString();
    }
    return this.bus.publish("president.approved", JSON.stringify(payload));
  }

  async grantBlanket(opts: BlanketOptions): Promise<string> {
    const payload: Record<string, unknown> = {
      category: opts.category,
      reason: opts.reason ?? "",
    };
    if (opts.expiresAt) {
      payload.expires_at = opts.expiresAt.toISOString();
    }
    if (opts.constraints) {
      payload.constraints = opts.constraints;
    }
    return this.bus.publish(
      "president.authorization.granted",
      JSON.stringify(payload),
    );
  }

  // revoke looks up the blanket approval's category from approvals_active so
  // callers only need the original approval event ID. Throws if the ID does
  // not correspond to an active blanket approval.
  async revoke(approvalEventId: string, reason: string): Promise<string> {
    const { rows } = await this.bus.pool.query<{ category: string | null }>(
      `SELECT category FROM approvals_active
        WHERE approval_event_id = $1 AND kind = 'blanket' LIMIT 1`,
      [approvalEventId],
    );
    const category = rows[0]?.category;
    if (!category) {
      throw new Error(
        `approvals: blanket approval ${approvalEventId} not found or has no category`,
      );
    }
    return this.bus.publish(
      "president.authorization.revoked",
      JSON.stringify({
        category,
        revoked_approval_id: approvalEventId,
        reason,
      }),
    );
  }
}
